### PR TITLE
Adjust widgets to support immutability

### PR DESCRIPTION
**Pull request by terabytesoftw**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  | https://github.com/yiisoft/yii-bulma/pull/20

### CODE REVIEW

**Comment by terabytesoftw** [comment]
We have two types of widgets:

Widget simple:

```php
$widget = MyWidget::Widget();
$widget->render();
```

Widget Begin/End:

```php
$widget = MyWidget::Begin();
$widget->start();
    // content
$widget->end();
```

**Comment by terabytesoftw** [comment]
Another variant but there will be a great BC:

We have two types of widgets:

Widget simple:

```php
$widget = MyWidget::widget();
$widget->render();
```

Widget begin/end:

```php
$widget = MyWidget::widget();
$widget->begin();
    // content
$widget->end();
```

**Comment by samdark** [comment]
Do you mean:

```php
$widget = MyWidget::widget()->option1(true)->option2(false);
$widget->begin();
    // content
$widget->end();
```

?

**Comment by terabytesoftw** [comment]
> Do you mean:
> 
> ```
> $widget = MyWidget::widget()->option1(true)->option2(false);
> $widget->begin();
>     // content
> $widget->end();
> ```
> 
> ?

yes.

**Comment by samdark** [comment]
In a template that would be:

```php
<?php $widget = MyWidget::widget()->option1(true)->option2(false) ?>
<?= $widget->begin() ?>
    Content
<?= $widget->end() ?>
```

It could also be:

```php
<?= MyWidget::widget()->options(['class' => 'testMe'])->begin() ?>
    // content
<?= MyWidget::end() ?>
```

**Comment by terabytesoftw** [comment]
> In a template that would be:
> 
> ```
> <?php $widget = MyWidget::widget()->option1(true)->option2(false) ?>
> <?= $widget->begin() ?>
>     Content
> <?= $widget->end() ?>
> ```
> 
> It could also be:
> 
> ```
> <?= MyWidget::widget()->options(['class' => 'testMe'])->begin() ?>
>     // content
> <?= MyWidget::end() ?>
> ```

yes, i have added the test.

File: tests/WidgetTest.php [hunk 1]
```diff
@@ -50,10 +84,10 @@ public function testStackTracking(): void
     public function testStackTrackingDisorder(): void
     {
         $this->expectException(InvalidConfigException::class);
-        TestWidgetA::begin();
-        TestWidgetB::begin();
-        TestWidgetA::end();
-        TestWidgetB::end();
+        $a = TestWidgetA::widget();
```
Comment (samdark) [inline tests/WidgetTest.php:87]:
```suggestion
        $a = TestWidgetA::widget()->begin();
```

File: tests/WidgetTest.php [hunk 2]
```diff
@@ -50,10 +84,10 @@ public function testStackTracking(): void
...
+        $b = TestWidgetB::widget();
```
Comment (samdark) [inline tests/WidgetTest.php:88]:
```suggestion
        $b = TestWidgetB::widget()->begin();
```

File: tests/WidgetTest.php [hunk 3]
```diff
@@ -29,19 +30,52 @@ public function testWidgetArrayConfig(): void
...
+
     /**
      * @depends testBeginEnd
      */
     public function testStackTracking(): void
     {
+        $widget = TestWidget::widget();
         $this->expectException(InvalidConfigException::class);
-        TestWidget::end();
+        $widget->end();
```
Comment (samdark) [inline tests/WidgetTest.php:78]:
```suggestion
        TestWidget::end();
```

File: tests/WidgetTest.php [hunk 4]
```diff
@@ -29,19 +30,52 @@ public function testWidgetArrayConfig(): void
...
+        $output = ImmutableWidget::end();
+
+        $this->assertSame('<run-new>', $output);
+    }
+
+    public function testStackTrackingWithImmutableWidget(): void
+    {
+        $widget = ImmutableWidget::widget();
+        $this->expectException(InvalidConfigException::class);
+        $widget->end();
```
Comment (samdark) [inline tests/WidgetTest.php:68]: `end()` is a static method.

File: tests/WidgetTest.php [hunk 5]
```diff
@@ -29,19 +30,52 @@ public function testWidgetArrayConfig(): void
...
+        $output = $widget->render();
+
+        $this->assertSame('<run-new>', $output);
+    }
+
+    public function testBeginEndWithImmutableWidget(): void
+    {
+        $widget = ImmutableWidget::widget()->id('new');
+        $widget->begin();
+        $output = $widget->end();
```
Comment (samdark) [inline tests/WidgetTest.php:51]: `end()` is a static method.

File: tests/WidgetTest.php [hunk 6]
```diff
@@ -29,19 +30,52 @@ public function testWidgetArrayConfig(): void
 
     public function testBeginEnd(): void
     {
-        TestWidgetA::begin()->id('test');
+        $widget = TestWidgetA::widget()->id('test')->begin();
```
Comment (samdark) [inline tests/WidgetTest.php:33]:
```suggestion
        TestWidgetA::widget()->id('test')->begin();
```

**Review by samdark — CHANGES_REQUESTED** [review]

File: tests/WidgetTest.php (hunk 6 above)
Comment (terabytesoftw) [reply tests/WidgetTest.php:33]: done.

File: tests/WidgetTest.php (hunk 5 above)
Comment (terabytesoftw) [reply tests/WidgetTest.php:51]: done.

File: tests/WidgetTest.php (hunk 4 above)
Comment (terabytesoftw) [reply tests/WidgetTest.php:68]: done.

File: tests/WidgetTest.php (hunk 3 above)
Comment (terabytesoftw) [reply tests/WidgetTest.php:78]: done.

File: tests/WidgetTest.php (hunk 2 above)
Comment (terabytesoftw) [reply tests/WidgetTest.php:88]: done.

File: tests/WidgetTest.php (hunk 1 above)
Comment (terabytesoftw) [reply tests/WidgetTest.php:87]: done.

**Review by samdark — APPROVED** [review]

**Comment by terabytesoftw** [comment]
Please check I think all the PRs that BC will do in the repositories are ready, are there any missing ?

**Comment by samdark** [comment]
- auth-client
- dataview
- yii-widgets

I'll fix these.

**Pull request merged**